### PR TITLE
[SPARK-40271][PYTHON] Support list type for `pyspark.sql.functions.lit`.

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR proposes to support `list` type for `pyspark.sql.functions.lit`.

### Why are the changes needed?

To improve the API usability.

### Does this PR introduce _any_ user-facing change?

Yes, now the `list` type is available for `pyspark.sql.functions.list` as below:

- Before
```python
>>> spark.range(3).withColumn("c", lit([1,2,3])).show()
Traceback (most recent call last):
...
: org.apache.spark.SparkRuntimeException: [UNSUPPORTED_FEATURE.LITERAL_TYPE] The feature is not supported: Literal for '[1, 2, 3]' of class java.util.ArrayList.
	at org.apache.spark.sql.errors.QueryExecutionErrors$.literalTypeUnsupportedError(QueryExecutionErrors.scala:302)
	at org.apache.spark.sql.catalyst.expressions.Literal$.apply(literals.scala:100)
	at org.apache.spark.sql.functions$.lit(functions.scala:125)
	at org.apache.spark.sql.functions.lit(functions.scala)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:577)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:374)
	at py4j.Gateway.invoke(Gateway.java:282)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:182)
	at py4j.ClientServerConnection.run(ClientServerConnection.java:106)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

- After
```python
>>> spark.range(3).withColumn("c", lit([1,2,3])).show()
+---+---------+
| id|        c|
+---+---------+
|  0|[1, 2, 3]|
|  1|[1, 2, 3]|
|  2|[1, 2, 3]|
+---+---------+
```

### How was this patch tested?

Added doctest & unit test.

### CODE REVIEW

**Comment by itholic** [comment]
cc @zhengruifeng

File: python/pyspark/sql/functions.py [hunk 1]
```diff
@@ -149,7 +149,17 @@ def lit(col: Any) -> Column:
     +------+---+
     |     5|  0|
     +------+---+
+
+    Support for list
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:153]:
```suggestion
    Create a literal from a list.
```

Comment (HyukjinKwon) [reply python/pyspark/sql/functions.py:153]: Can you fix "Python primitive type." in the docstring above?

File: python/pyspark/sql/tests/test_functions.py [hunk 2]
```diff
@@ -962,6 +962,11 @@ def test_lit_day_time_interval(self):
         actual = self.spark.range(1).select(lit(td)).first()[0]
         self.assertEqual(actual, td)
 
+    def test_lit_list(self):
+        test_list = [1, 2, 3]
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_functions.py:966]: Let's add a docstring like


```
# SPARK-12345: a short description of the test
```

See https://spark.apache.org/contributing.html

Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_functions.py:966]: Should probably have one more test with other fixed types. e.g.) `["a", 1, None, 1.0]`

File: python/pyspark/sql/functions.py (hunk 1 above)
Comment (HyukjinKwon) [reply python/pyspark/sql/functions.py:153]: Let's also add `versionchanged` directive for the list support.

**Review by HyukjinKwon — APPROVED** [review]
LGTM otherwise

File: python/pyspark/sql/tests/test_functions.py (hunk 2 above)
Comment (zhengruifeng) [inline python/pyspark/sql/tests/test_functions.py:966]: I think we can also add a test case like `[[1,2,3], [3,4]]`

**Comment by itholic** [comment]
Thanks, @HyukjinKwon and @zhengruifeng for the review!

**Review by zhengruifeng — APPROVED** [review]

File: python/pyspark/sql/functions.py [hunk 3]
```diff
@@ -149,7 +152,17 @@ def lit(col: Any) -> Column:
...
+    Create a literal from a list.
+
+    >>> spark.range(1).select(F.lit([1, 2, 3])).show()
+    +--------------+
+    |array(1, 2, 3)|
+    +--------------+
+    |     [1, 2, 3]|
+    +--------------+
     """
+    col = array(*[lit(item) for item in col]) if isinstance(col, list) else col
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:165]: Actually we should probably block the case of `[Column, Column]` because Scala side does not allow that case.

Comment (zhengruifeng) [reply python/pyspark/sql/functions.py:165]: so may looks like this:
```
    if isinstance(col, Column):
        return col
    elif isinstance(col, list):
        if any(isinstance(c, Column) for c in col):
            raise ValueError(...)
        return array(*[lit(item) for item in col])
    else:
        return _invoke_function("lit", col)
```

Comment (itholic) [reply python/pyspark/sql/functions.py:165]: qq: Maybe seems like the Scala side even doesn't allow for the single `Column` case ??

```scala
scala> df.select(lit(df.id)).show()
<console>:24: error: value id in class Dataset cannot be accessed in org.apache.spark.sql.Dataset[Long]
       df.select(lit(df.id)).show()
```

which is allowed in Python side:

```python
>>> df.select(F.lit(df.id)).show()
+---+
| id|
+---+
|  0|
|  1|
|  2|
|  3|
|  4|
|  5|
|  6|
|  7|
|  8|
|  9|
+---+
```

Comment (HyukjinKwon) [reply python/pyspark/sql/functions.py:165]: df.id isn't a way to access a column in scala. try either $"id" or df.col("id")

Comment (itholic) [reply python/pyspark/sql/functions.py:165]: Thanks!

File: python/pyspark/sql/functions.py [hunk 4]
```diff
@@ -155,15 +155,21 @@ def lit(col: Any) -> Column:
...
     |     [1, 2, 3]|
     +--------------+
     """
-    col = array(*[lit(item) for item in col]) if isinstance(col, list) else col
-    return col if isinstance(col, Column) else _invoke_function("lit", col)
+    if isinstance(col, Column):
+        return col
+    elif isinstance(col, list):
+        if any(isinstance(c, Column) for c in col):
+            raise ValueError("lit does not allow for list of Columns")
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:169]:
```suggestion
            raise ValueError("lit does not allow a column in a list")
```

File: python/pyspark/sql/tests/test_functions.py [hunk 5]
```diff
@@ -984,6 +984,10 @@ def test_lit_list(self):
         actual = self.spark.range(1).select(lit(test_list)).first()[0]
         self.assertEqual(actual, expected)
 
+        df = self.spark.range(10)
+        with self.assertRaisesRegex(ValueError, "lit does not allow for list of Columns"):
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_functions.py:988]:
```suggestion
        with self.assertRaisesRegex(ValueError, "lit does not allow a column in a list"):
```

**Review by HyukjinKwon — APPROVED** [review]
LGTM!

**Comment by zhengruifeng** [comment]
Merged to master, thank you!

**Comment by zhengruifeng** [comment]
oops, sorry that @HyukjinKwon 's suggestions don't get committed. Would you @itholic mind updating it in a followup? Or I can do that too.

**Comment by itholic** [comment]
@zhengruifeng Let me address it. Thanks! :-)

**Pull request closed without merging**